File: Binary Search/Python/Search a 2D Matrix.py
// brute TC:O(rows * cols) SC:O(1)
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows = len(matrix)
        cols = len(matrix[0])
        for i in range(rows):
            for j in range(cols):
                if matrix[i][j] == target:return True
        return False

## optimal TC:O(log(rows*cols)) SC:(1)
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows = len(matrix)
        cols = len(matrix[0])
        low = 0
        high = rows * cols-1
        while low<=high:
            mid = low+(high-low)//2
            row=mid//cols
            col = mid%cols
            if matrix[row][col] == target:return True
            if matrix[row][col]<target: low=mid+1
            else:high=mid-1
        return False

File: Binary Search/Search a 2D Matrix.cpp
// brute TC:O(rows * cols)  SC:(1)
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int rows = matrix.size();
        int cols = matrix[0].size();
        for(int i=0;i<rows;i++){
            for(int j=0;j<cols;j++){
                if(matrix[i][j] == target) return true;
            }
        }
        return false;
    }
};

// optimal  TC: O(log(rows * cols)) , SC: O(1)
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int rows = matrix.size();
        int cols = matrix[0].size();
        int low = 0, high = rows * cols - 1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            int row = mid / cols;
            int col = mid % cols;
            if (matrix[row][col] == target)
                return true;
            else if (matrix[row][col] < target)
                low = mid + 1;
            else
                high = mid - 1;
        }
        return false;
    }
};
